sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt

```diff
@@ -1,65 +1,436 @@
+            }
+        }
+    }
+
+    private fun configurePurchaseInfo(skuDetails: Map<String, SkuDetails>, purchases: List<Purchase>): List<com.qonversion.android.sdk.entity.Purchase> {
+        val result = mutableListOf<com.qonversion.android.sdk.entity.Purchase>()
+
+        purchases.forEach {
+            val skuDetail = skuDetails[it.sku]
+            if (skuDetail != null) {
+                val purchaseInfo = Pair.create(skuDetail, it)
+                val purchase = converter.convert(purchaseInfo)
+                result.add(purchase)
+            }
+        }
+
+        return result
+    }
+
+    private fun configureSkuDetails(skuDetails: List<SkuDetails>): Map<String, SkuDetails> {
+        val formattedData = mutableMapOf<String, SkuDetails>()
+        skuDetails.forEach {
+            formattedData[it.sku] = it
+        }
+
+        return formattedData
+    }
+
+    private fun getInstallDate(): Long {
+        if (installDate > 0) {
+            return installDate
+        }
+
+        installDate = context.packageManager.getPackageInfo(context.packageName, 0).firstInstallTime.milliSecondsToSeconds()
+
+        return installDate
+    }
+
+    private fun continueLaunchWithPurchasesInfo(advertisingId: String? = null, callback: QonversionLaunchCallback?) {
+        val installDate = getInstallDate()
+        billingService.queryPurchases(
+            onQueryCompleted = { purchases ->
+                if (purchases.isEmpty()) {
+                    repository.init(installDate = installDate, idfa = advertisingId, callback = callback)
+                    return@queryPurchases
+                }
+
+                billingService.getSkuDetailsFromPurchases(
+                    purchases,
+                    onCompleted = { skuDetails ->
+                        val formattedSkuDetails: Map<String, SkuDetails> = configureSkuDetails(skuDetails)
+                        val purchasesInfo = configurePurchaseInfo(formattedSkuDetails, purchases)
+                        repository.init(installDate, advertisingId, purchasesInfo, callback)
+                    },
+                    onFailed = {
+                        repository.init(installDate = installDate, idfa = advertisingId, callback = callback)
+                    })
+            },
+            onQueryFailed = {
+                repository.init(installDate = installDate, idfa = advertisingId, callback = callback)
+            })
+    }
+
+    private fun getLaunchCallback(callback: QonversionLaunchCallback?): QonversionLaunchCallback {
+        return object: QonversionLaunchCallback {
+            override fun onSuccess(launchResult: QLaunchResult) {
+                isLaunchingFinished = true
+                this@QProductCenterManager.launchResult = launchResult
+                launchError = null
+
+                loadStoreProductsIfPossible(launchResult)
+
+                executePermissionsBlock()
+
+                callback?.onSuccess(launchResult)
+            }
+
```

Comment:
Maybe, we should delete these blank lines or just do a uniform style in onSuccess and onError callbacks